File: penguin-popper/orig/graphic.java
public class graphic{
		public class point{
			float x, y;
			public point(float xa, float ya){
				x=xa;
				y=ya;
			}
		}
		public class poly{
			point[] point;
			boolean varColor=false;
			java.awt.Color color;
			public void draw(java.awt.Graphics2D g2, double xc, double yc, double scale, double angle, boolean flip){
				if (point.length>2){
					int[] xcoords=new int[point.length], ycoords=new int[point.length];
					for (int i=0; i<point.length; i++){
						double d=dist(point[i].x, point[i].y, 400, 285)/285*scale, a=getAngle(point[i].x, point[i].y, 400, 285)+angle;
						xcoords[i]=(int)(xc-d*Math.cos(a));
						if (flip) xcoords[i]=(int)(2*xc-xcoords[i]);
						ycoords[i]=(int)(yc-d*Math.sin(a));
					}
					g2.setColor(color);
					g2.fillPolygon(xcoords, ycoords, point.length);
				}
			}
			public void draw(java.awt.Graphics2D g2, double xc, double yc, double scale, double angle, boolean flip, java.awt.Color vColor){
				if (point.length>2){
					int[] xcoords=new int[point.length], ycoords=new int[point.length];
					for (int i=0; i<point.length; i++){
						double d=dist(point[i].x, point[i].y, 400, 285)/285*scale, a=getAngle(point[i].x, point[i].y, 400, 285)+angle;
						xcoords[i]=(int)(xc-d*Math.cos(a));
						if (flip) xcoords[i]=(int)(2*xc-xcoords[i]);
						ycoords[i]=(int)(yc-d*Math.sin(a));
					}
					if (varColor) g2.setColor(vColor);
					else g2.setColor(color);
					g2.fillPolygon(xcoords, ycoords, point.length);
				}
			}
			public boolean inside(int x, int y){
				boolean odd=false;
				if (point.length>2){
					int j=point.length-1;
					for (int i=0; i<point.length; i++){
						if (point[i].y<y && point[j].y>=y || point[j].y<y && point[i].y>=y){
							if (point[i].x+(y-point[i].y)/(point[j].y-point[i].y)*(point[j].x-point[i].x)<x) odd=!odd;
						}
						j=i;
					}
				}
				return odd;
			}
			public void scale (double scale){
				for (int i=0; i<point.length; i++){
					double d=dist(point[i].x, point[i].y, 400, 285)/285*scale, a=getAngle(point[i].x, point[i].y, 400, 285);
					point[i].x=(float)(400-d*Math.cos(a));
					point[i].y=(float)(285-d*Math.sin(a));
				}
			}
			public void drawScaled(java.awt.Graphics2D g2, double xc, double yc){
				int[] xcoords=new int[point.length], ycoords=new int[point.length];
				for (int i=0; i<point.length; i++){
					xcoords[i]=(int)(xc-400+point[i].x);
					ycoords[i]=(int)(yc-285+point[i].y);
				}
				g2.setColor(color);
				g2.fillPolygon(xcoords, ycoords, point.length);
			}
		}
		poly[] poly;
		public graphic (String input){
			int polyIndex=0, pointIndex=0;
			String current="", mode="index";
			float x=0;
			for (int i=0; i<input.length()-1; i++){
				if (input.charAt(i)==' '){
					if (mode=="index"){
						poly=new poly[java.lang.Integer.parseInt(current)];
						mode="color";
					} else if (mode=="color"){
						poly[polyIndex]=new poly();
						if (current.charAt(0)=='v') poly[polyIndex].varColor=true;
						else poly[polyIndex].color=new java.awt.Color(java.lang.Integer.parseInt(current));
						polyIndex++;
						mode="points";
					} else if (mode=="points"){
						poly[polyIndex-1].point=new point[java.lang.Integer.parseInt(current)];
						pointIndex=0;
						mode="x";
					} else if (mode=="x"){
						x=java.lang.Float.parseFloat(current);
						mode="y";
					} else if (mode=="y"){
						poly[polyIndex-1].point[pointIndex]=new point(x, java.lang.Float.parseFloat(current));
						pointIndex++;
						mode="x";
					}
					current="";
				} else if (input.charAt(i)==','){
					mode="color";
					current="";
				} else current+=input.charAt(i);
			}
		}
		public void draw(java.awt.Graphics2D g2, double xc, double yc, double scale, double angle, boolean flip){
			for (int i=0; i<poly.length; i++) poly[i].draw(g2, xc, yc, scale, angle, flip);
		}
		public void draw(java.awt.Graphics2D g2, double xc, double yc, double scale, double angle, boolean flip, java.awt.Color vColor){
			for (int i=0; i<poly.length; i++) poly[i].draw(g2, xc, yc, scale, angle, flip, vColor);
		}
		public void drawScaled(java.awt.Graphics2D g2, double xc, double yc){
			for (int i=0; i<poly.length; i++) poly[i].drawScaled(g2, xc, yc);
		}
		public double getAngle(double x1, double y1, double x2, double y2){
			return x2<x1 ? Math.atan((y2-y1)/(x2-x1))+3.1415926536 : Math.atan((y2-y1)/(x2-x1));
		}
		public double dist (double x1, double y1, double x2, double y2){
			return Math.sqrt((x1-x2)*(x1-x2)+(y1-y2)*(y1-y2));
		}
		public void scale(double scale){
			for (int i=0; i<poly.length; i++) poly[i].scale(scale);
		}
	}

File: penguin-popper/graphic.js
const dist = (x1, y1, x2, y2) =>
  Math.sqrt((x1 - x2) * (x1 - x2) + (y1 - y2) * (y1 - y2));

const getAngle = (x1, y1, x2, y2) =>
  x2 < x1
    ? Math.atan((y2 - y1) / (x2 - x1)) + 3.1415926536
    : Math.atan((y2 - y1) / (x2 - x1));

const fillPolygon = (ctx, xcoords, ycoords, len) => {
  ctx.beginPath();
  for (let i = 0; i < len; i++) {
    ctx.lineTo(xcoords[i], ycoords[i]);
  }
  ctx.fill();
};

const toColor = n =>
  isNaN(n) ? n : '#' + ('000000' + (n >>> 0).toString(16)).slice(-6);

class Poly {
  draw(ctx, xc, yc, scale, angle, flip, vColor) {
    if (this.point.length > 2) {
      let xcoords = [],
        ycoords = [];
      for (let i = 0; i < this.point.length; i++) {
        let d =
            (dist(this.point[i].x, this.point[i].y, 400, 285) / 285) * scale,
          a = getAngle(this.point[i].x, this.point[i].y, 400, 285) + angle;
        xcoords[i] = xc - d * Math.cos(a);
        if (flip) xcoords[i] = 2 * xc - xcoords[i];
        ycoords[i] = yc - d * Math.sin(a);
      }
      if (this.varColor) ctx.fillStyle = toColor(vColor);
      else ctx.fillStyle = toColor(this.color);
      fillPolygon(ctx, xcoords, ycoords, this.point.length);
    }
  }
  inside(x, y) {
    let odd = false;
    if (this.point.length > 2) {
      let j = this.point.length - 1;
      for (let i = 0; i < this.point.length; i++) {
        if (
          (this.point[i].y < y && this.point[j].y >= y) ||
          (this.point[j].y < y && this.point[i].y >= y)
        ) {
          if (
            this.point[i].x +
              ((y - this.point[i].y) / (this.point[j].y - this.point[i].y)) *
                (this.point[j].x - this.point[i].x) <
            x
          )
            odd = !odd;
        }
        j = i;
      }
    }
    return odd;
  }
  scale(scale) {
    for (let i = 0; i < this.point.length; i++) {
      let d = (dist(this.point[i].x, this.point[i].y, 400, 285) / 285) * scale,
        a = getAngle(this.point[i].x, this.point[i].y, 400, 285);
      this.point[i].x = 400 - d * Math.cos(a);
      this.point[i].y = 285 - d * Math.sin(a);
    }
  }
  drawScaled(ctx, xc, yc) {
    let xcoords = [],
      ycoords = [];
    for (let i = 0; i < this.point.length; i++) {
      xcoords[i] = xc - 400 + this.point[i].x;
      ycoords[i] = yc - 285 + this.point[i].y;
    }
    ctx.fillStyle = toColor(this.color);
    fillPolygon(ctx, xcoords, ycoords, this.point.length);
  }
}

export class graphic {
  constructor(input) {
    this.poly = [];
    let polyIndex = 0,
      pointIndex = 0;
    let current = '',
      mode = 'index';
    let x = 0;
    for (let i = 0; i < input.length; i++) {
      if (input.charAt(i) == ' ') {
        if (mode == 'index') {
          this.poly = [];
          mode = 'color';
        } else if (mode == 'color') {
          this.poly[polyIndex] = new Poly();
          if (current.charAt(0) == 'v') this.poly[polyIndex].varColor = true;
          else this.poly[polyIndex].color = current;
          polyIndex++;
          mode = 'points';
        } else if (mode == 'points') {
          this.poly[polyIndex - 1].point = [];
          pointIndex = 0;
          mode = 'x';
        } else if (mode == 'x') {
          x = parseFloat(current);
          mode = 'y';
        } else if (mode == 'y') {
          this.poly[polyIndex - 1].point[pointIndex] = {
            x,
            y: parseFloat(current)
          };
          pointIndex++;
          mode = 'x';
        }
        current = '';
      } else if (input.charAt(i) == ',') {
        mode = 'color';
        current = '';
      } else current += input.charAt(i);
    }
  }
  draw(ctx, xc, yc, scale, angle, flip, vColor) {
    for (let i = 0; i < this.poly.length; i++)
      this.poly[i].draw(ctx, xc, yc, scale, angle, flip, vColor);
  }
  drawScaled(ctx, xc, yc) {
    for (let i = 0; i < this.poly.length; i++)
      this.poly[i].drawScaled(ctx, xc, yc);
  }
  scale(scale) {
    for (let i = 0; i < this.poly.length; i++) this.poly[i].scale(scale);
  }
}
